File: clients/dotnet/src/Varta.Client/DecodeError.cs
namespace Varta;

/// <summary>
/// Discriminator for <see cref="DecodeError"/> failures returned by
/// <see cref="Frame.Decode"/>.
/// </summary>
public enum DecodeErrorKind
{
    BadMagic,
    BadVersion,
    BadCrc,
    BadStatus,
    StallOnWire,
    BadPid,
    BadTimestamp,
    BadNonce,
}

/// <summary>
/// Thrown by <see cref="Frame.Decode"/> when a wire buffer fails any of
/// the VLP v0.2 structural checks. Validation precedence follows the
/// spec: magic → version → CRC → status → StallOnWire → pid →
/// timestamp → nonce.
/// </summary>
public sealed class DecodeError : Exception
{
    public DecodeErrorKind Kind { get; }
    public string Detail { get; }

    public DecodeError(DecodeErrorKind kind, string detail)
        : base($"varta: decode failed: {kind} ({detail})")
    {
        Kind = kind;
        Detail = detail;
    }
}

File: clients/java/lib/src/main/java/health/varta/DecodeError.java
package health.varta;

/**
 * Thrown by {@link Frame#decode(byte[])} / {@link Frame#decode(java.nio.ByteBuffer)}
 * when a VLP frame fails validation. Inspect {@link #kind()} for the
 * structured rejection reason.
 */
public final class DecodeError extends RuntimeException {
    private static final long serialVersionUID = 1L;

    private final DecodeErrorKind kind;

    public DecodeError(DecodeErrorKind kind, String detail) {
        super(kind.name() + ": " + detail);
        this.kind = kind;
    }

    public DecodeErrorKind kind() {
        return kind;
    }
}
